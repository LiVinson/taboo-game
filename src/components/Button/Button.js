import React from "react"
import "./Button.scss"


const Button = ({text, handleClick, type}) => {


  return <button className={`button ${type ? "button--"+type: ""}`} onClick={handleClick}>{text}</button>
}
export default Button
