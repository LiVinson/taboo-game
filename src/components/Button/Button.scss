.button {
    border:none;
    color: var(--color-primary);
    background-color: var(--color-off-white);
    font-size:2rem;
    width:47%;
    padding: 1rem 0rem;


    &--lg {

    }

    &--primary-color {
        color: var(--color-off-white);
        background-color: var(--color-primary--light);


        &--lg {

        }
    }


  

}