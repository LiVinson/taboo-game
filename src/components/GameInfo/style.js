import styled from "styled-components"

export const StyledGameInfo = styled.div`
   display: flex;
   flex-wrap: wrap;
  justify-content: space-between;
  margin: 0 auto;

  & > * {
    margin-bottom:1.5rem;
    display: ${props => props.open && "block"};
  }
`

StyledGameInfo.displayName = "StyledGameInfo"
