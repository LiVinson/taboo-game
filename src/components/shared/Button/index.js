import React from "react"
import { StyledButton, StyledPrimaryButton, StyledLargeButton, StyledTextButton} from "./style"

//Takes in object of properties passed down. Typically includes type, onClick, text, and any other required attributes
export const Button = (props) => {
  return <StyledButton {...props}>{props.text}</StyledButton>
}


export const PrimaryButton = (props) => {
  return <StyledPrimaryButton {...props}>{props.text}</StyledPrimaryButton>
}

export const LargeButton = (props) => {
  return <StyledLargeButton {...props}>{props.text}</StyledLargeButton>
}

export const TextButton = (props) => {
  return <StyledTextButton {...props}>{props.text}</StyledTextButton>
}