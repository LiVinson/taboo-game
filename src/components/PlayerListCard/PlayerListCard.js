import React from 'react'
import PropTypes from 'prop-types'
import { ButtonTabooCard, TabooCard } from 'components/shared/TabooCard'

const PlayerListCard = ({ buttonInfo, tabooWord, children }) => {
	return buttonInfo ? (
		<ButtonTabooCard tabooWord={tabooWord} buttons={buttonInfo}>
			{children}
		</ButtonTabooCard>
	) : (
		<TabooCard tabooWord={tabooWord}>{children}</TabooCard>
	)
}

PlayerListCard.defaultProps = {
	tabooWord: "Players"
}

PlayerListCard.propTypes = {
	buttonInfo: PropTypes.array,
	tabooWord: PropTypes.string.isRequired,
	children: PropTypes.node.isRequired

}

export default PlayerListCard
