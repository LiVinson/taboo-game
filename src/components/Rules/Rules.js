import React from 'react'
import { Redirect } from 'react-router-dom'
import PropTypes from 'prop-types'
import { ButtonTabooCard } from 'components/shared/TabooCard'
import rules from './rulesText'
import { RulesTitle, RulesText } from './style'

/*
  Can be displayed in 2 ways: on it's own route or within the play route.
  If on home route closing instructions goes back to /home. If on play route,
  calls a callback provided from GameInfo that will toggle Rules closed.
*/
export default class Rules extends React.Component {
	constructor(props) {
		super(props)
		this.state = {
			currentRule: 0,
		}
	}

	//When user clicks Back. If already showing first rule, either call toggleGameInfo (during game play)
	//or return to home route. Otherwise set state to previous rule
	displayPreviousRule = (event) => {
		const { currentRule } = this.state
		if (currentRule > 0) {
			this.setState((state) => {
				return {
					currentRule: state.currentRule - 1,
				}
			})
		} else {
			if (this.props.toggleGameInfo) {
				this.props.toggleGameInfo(event)
			} else {
				this.props.history.push('/home')
			}
		}
	}

	//When user clicks Next. If already showing last rule, either call toggleGameInfo (during game play)
	//or return to home route. Otherwise set state to next rule
	displayNextRule = (event) => {
		const { currentRule } = this.state
		if (currentRule < rules.length - 1) {
			this.setState((state) => {
				return {
					currentRule: state.currentRule + 1,
				}
			})
		} else {
			if (this.props.toggleGameInfo) {
				this.props.toggleGameInfo(event)
			} else {
				this.props.history.push('/home')
			}
		}
	}

	render() {
		const { currentRule } = this.state
		const buttonInfo = [
			{
				text: currentRule > 0 ? 'Back' : 'Close',
				onClick: this.displayPreviousRule,
				name: this.props.toggleGameInfo && 'showRules',
				type: 'button',
			},
			{
				text: currentRule < rules.length - 1 ? 'Next' : 'Close',
				onClick: this.displayNextRule,
				name: this.props.toggleGameInfo && 'showRules',
				type: 'button',
			},
		]

		return (
			<ButtonTabooCard tabooWord="How to Play" buttons={buttonInfo} type="home">
				<DisplayRulesText topicId={currentRule} />
			</ButtonTabooCard>
		)
	}
}

Rules.propTypes = {
	history: PropTypes.object,
	toggleGameInfo: PropTypes.func,
}

//Loops over imported array of rules and displays title and paragraphs for rule displayed
const DisplayRulesText = ({ topicId }) => {
	const rulesInfo = rules[topicId]
	if (rulesInfo) {
		return (
			<React.Fragment>
				<RulesTitle>{rulesInfo.title}</RulesTitle>
				{rulesInfo.text.map((paragraph, index) => (
					<RulesText key={index}>{paragraph}</RulesText>
				))}
			</React.Fragment>
		)
	} else {
		//Shouldn't be possible, but just in case.
		return <Redirect to="/404" />
	}
}

DisplayRulesText.propTypes = {
	topicId: PropTypes.number.isRequired,
}
