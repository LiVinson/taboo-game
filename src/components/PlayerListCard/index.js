import PlayerListCard from "./PlayerListCard"

export default PlayerListCard