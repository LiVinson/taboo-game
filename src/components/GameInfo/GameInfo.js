import React from 'react'
import PropTypes from "prop-types"
import Rules from 'components/Rules'
import PlayerListCard from 'components/PlayerListCard'
import { FilteredTabooList } from 'components/shared/TabooCard'
import { Button } from 'components/shared/Button'
import { StyledGameInfo } from './style'

//Displays a link for Show Rules and Show Teams. On click, displays a TabooCard with Rules/Team 
//and link changes to Hide Rules/Teams
class GameInfo extends React.Component {
	constructor(props) {
		super(props)
		this.state = {
			showRules: false,
			showTeams: false,
		}
	}
	//Ttriggered on click of Show Rules or Show Teams button. Updates state property based on which button clicked
	toggleGameInfo = (event) => {
		const gameInfo = event.target.name
		this.setState((state) => {
			return {
				[gameInfo]: !state[gameInfo],
			}
		})
	}

	render() {
		const teams = ['team 1', 'team 2']

		const buttonInfo = [
			{
				text: this.state.showRules ? 'Hide Rules' : 'Show Rules',
				type: 'button',
				onClick: this.toggleGameInfo,
				name: 'showRules',
				border: true,
			},
			{
				text: this.state.showTeams ? 'Hide Teams' : 'Show Teams',
				type: 'button',
				onClick: this.toggleGameInfo,
				name: 'showTeams',
				border: true,
			},
		]

		const { showRules, showTeams } = this.state
		
		//Displays 2 buttons and 2 Taboo cards containing Rules and list team list when showing or nothing when hidden
		return (
			<StyledGameInfo open={showRules || showTeams}>
				{<Button {...buttonInfo[0]} />}
				{<Button {...buttonInfo[1]} />}
				{showRules && <Rules toggleGameInfo={this.toggleGameInfo} />}
				{showTeams && (
					<PlayerListCard>
						{teams.map((team) => (
							<FilteredTabooList
								key={team}
								unfilteredList={this.props.players}
								filterKey="team"
								filterValue={team}
								displayProperty="name"
								listTitle={team}
								specialKey="playerId"
								specialValue={this.props.currentPlayer.playerId}
							/>
						))}
					</PlayerListCard>
				)}
			</StyledGameInfo>
		)
	}
}

GameInfo.propTypes = {
	players: PropTypes.array.isRequired,
	currentPlayer: PropTypes.object.isRequired
}

export default GameInfo
