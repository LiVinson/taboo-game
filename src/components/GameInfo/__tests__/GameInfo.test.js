import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { ThemeProvider } from 'styled-components'
import theme from '../../../global-design/theme'
import GameInfo from '../GameInfo'
import Rules from 'components/Rules'
import PlayerListCard from 'components/PlayerListCard'
import { Button } from 'components/shared/Button'

describe('Game Info rendering and functionality', () => {
	const defaultProps = {
		players: [
			{ name: 'Player 1', playerId: '12345', team: 'team 1' },
			{ name: 'Player 2', playerId: '23456', team: 'team 1' },
			{ name: 'Player 3', playerId: '34567', team: 'team 2' },
			{ name: 'Player 4', playerId: '45678', team: 'team 2' },
		],
		currentPlayer: { name: 'Player 1', playerId: '12345', team: 'team 1' },
    }
	
	//Used to similate the event created from button click
    const mockEvent = {
        target: {
            name: ""
        }
    }

	test('renders correctly', () => {
		const props = {
			...defaultProps,
		}
		const wrapper = renderer
			.create(
				<ThemeProvider theme={theme}>
					<GameInfo {...props} />
				</ThemeProvider>
			)
			.toJSON()

		expect(wrapper).toMatchSnapshot()
	})

	test('contains 2 Buttons and no Rules or PlayerListCard components by default', () => {
		const props = {
			...defaultProps,
		}
		const wrapper = shallow(
				<GameInfo {...props} />
		)

		expect(wrapper.dive().find(Button)).toHaveLength(2)
		expect(wrapper.dive().find(Rules)).toHaveLength(0)
		expect(wrapper.dive().find(PlayerListCard)).toHaveLength(0)
	})

	test('Rules and PlayerListCard components render when Show buttons clicked', () => {
		const props = {
			...defaultProps,
		}
		const wrapper = shallow(
				<GameInfo {...props} />
		)

        //Call method that is called when button is clicked. Calling method directly instead of click to avoid using mount()
        mockEvent.target.name = "showRules"
        wrapper.instance().toggleGameInfo(mockEvent)
        
        mockEvent.target.name = "showTeams"
        wrapper.instance().toggleGameInfo(mockEvent)
     
		expect(wrapper.dive().find(Rules)).toHaveLength(1)
		expect(wrapper.dive().find(PlayerListCard)).toHaveLength(1)      
	})
})
