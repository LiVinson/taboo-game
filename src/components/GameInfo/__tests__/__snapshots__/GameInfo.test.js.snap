// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game Info rendering and functionality renders correctly 1`] = `
<div
  className="sc-fzqNJr leJyqT"
  open={false}
>
  <button
    className="sc-AxmLO gqfOdi"
    name="showRules"
    onClick={[Function]}
    type="button"
  >
    Show Rules
  </button>
  <button
    className="sc-AxmLO gqfOdi"
    name="showTeams"
    onClick={[Function]}
    type="button"
  >
    Show Teams
  </button>
</div>
`;
